oeps/oep-0007.rst

```diff
@@ -0,0 +1,462 @@
+
+
+Rationale
+=========
+
+    * Python 2 will only receive support from the Python core devs until 2020.
+    * Django will only support Python 2 until 1.11LTS (which is also supported
+      until 2020).
+    * New features are being made available on Python 3 first, and are only
+      sometimes backported to Python 2.
+    * New syntax extensions (`async`/`await`, `yield from`, and the `@`
+      operator, for instance) are only being introduced in Python 3.
+    * Python 3 provides better separation between text and binary data.
+    * Python 3 provides better memory usage through a more thorough use of
+      iterators and smart collections.
+    * Major language performance work is being done on Python 3 only.
+    * Almost all major libraries now support Python 3.
+    * We can approach the problem now at a relaxed pace, or deal with it in a
+      panic later, diverting valuable resources from other important work.
+    * The longer we wait, the more code we will produce that needs to be
+      migrated.
+    * Using Python 3 gives us a strong story as a leader in the Python and open
+      source communities.
+    * Python 3 makes for more compelling job postings [Citation needed]
+
+
+Scope
+=====
+
+There are several areas of work that need to be done
+
+* Setting policy
+* Changing coding practice.
+* Migrating existing code.
+* Deploying edx.org services with Python 3.
+* Deprecating Python 2 support.
+
+These areas have certain dependencies between them, e.g., we cannot deploy a
+service under Python 3 until all of its code has been migrated.  However, this
+should not be taken as a chronological list of tasks.  We may begin deprecating
+Python 2 support on certain repositories before existing code in other
+repositories has been migrated.
+
+
+Choosing a Python version
++++++++++++++++++++++++++
+
+All new Python projects should be written using Python 3 unless a compelling
+reason (such as incompatible support libraries with no reasonable alternative
+available) necessitates that we stick with Python 2.
+
+
+New libraries should additionally support Python 2 using a single code-base
+with six as a compatibility layer if they will be used by projects that run on
+Python 2.  Open source libraries we maintain that are generally useful beyond
+our services should also maintain compatibility with Python 2 until support for
+Python 2 is dropped by the core developers, which is scheduled to happen in
+April 2020.
+
+Deployable projects only need to support a single version of Python.  Libraries
+will often need to support a range of versions.  Libraries written in Python 3
+should initially support versions 3.5 and up.  Libraries that support Python 2
+should only support version 2.7.  Future version deprecations are outside the
+scope of this document.
+
+The end goal is for all of our services to be deployed under Python 3 or
+replaced by newer services before the April 2020 end of support.
+
+
+Migrating code
+++++++++++++++
+
```

Comment:
We should have links to six and future's documentation somewhere in this document